hypothesis-python/src/hypothesis/strategies/_internal/core.py

```diff
@@ -1204,7 +1214,22 @@ def defer_recursion(thing, producer):
     # may be able to fall back on type annotations.
     if issubclass(thing, enum.Enum):
         return sampled_from(thing)
-
+    # Handle types resolvable by extras
+    try:
+        from hypothesis.extra.numpy import from_type_internal
+
+        return from_type_internal(thing)  # pragma: no cover
+    except ImportError:
+        if getattr(thing, "__module__", None) == "numpy":  # pragma: no cover
```

Comment:
Ah, this is a bit tricky: for performance reasons, we don't want to cause Numpy to be imported even if it's installed.

Fortunately, this usually isn't much of a problem: we can check `if "numpy" in sys.modules and thing.__module__ == getattr(thing, "__module__", None) == "numpy":` - if Numpy _hasn't_ been imported already (in sys.modules), then we can be confident that the thing we're resolving is not from Numpy.

Without that, I think this would also be a bit _too_ general, in that `np.dtype(thing)` might work for more things than we really want, here!

----------

I'd probably also inline the function you wrote in the Numpy extra here, since we don't want it separately callable.  Sketch:

```python
np = sys.modules.get("numpy")
if np and thing.__module__ == getattr(thing, "__module__", None) == "numpy":
    try:
        dtype = np.dtype(thing)
    except Exception:
        pass
    else:
        from hypothesis.extra.numpy import from_dtype

        return from_dtype(dtype)
```

Reply:
Good call on the importing business, I didn't consider performance and numpy must indeed already be loaded.

Now, for the inline-or-separately -callable question, let me explain a reason for putting it in `extra.numpy`. Your call in the end, but maybe I'll convince you.

I mentioned that I found the `ndarray` resolution:

https://github.com/HypothesisWorks/hypothesis/blob/8cce2dbf733bba78257b4d3fe3ac5c42b9a43820/hypothesis-python/src/hypothesis/strategies/_internal/types.py#L602

This is not sufficient to support parameterized arrays (`np.ndarray[int]` or `npt.NDArray[int]`, in the future also shapes). That will require real logic, not only a lookup table. So, `ndarray` resolution will also end up here.

I'm not planning to put that into this PR, but scope creep is real you know :-).

This will end up being a lot of numpy-specific logic to put into core, and then maybe same for pandas, and so on.

For this reason, my choice would be to put it into extras.numpy somehow. Maybe it would be better if the method was underscored, or moved into separate `extras.numpy.internal` namespace?

Reply:
Let's try to ship the simplest thing we can, here.

https://github.com/HypothesisWorks/hypothesis/issues/3150 is a separate issue, and we can deal with those complications in a dedicated PR *after* we've shipped the valuable stuff in this one.